File: Unit02/clock.py
# Alex Tedesco

class Clock:
    __slots__ = ["__hours", "__minutes", "__seconds", "__am_or_pm"]
    def __init__(self, hours=12, minutes=0, seconds=1, am_or_pm="AM"):
        self.__hours = hours
        self.__minutes = minutes
        self.__seconds = seconds
        self.__am_or_pm = am_or_pm

    def get_hours (self):
        return self.__hours

    def get_minutes (self):
        return self.__minutes

    def get_seconds (self):
        return self.__seconds
    
    def get_am_or_pm (self):
        return self.__am_or_pm

    def __str__ (self):
        # <hours>:<minutes>:<seconds> *M
        return str (self.__hours) + ":" + f'{self.__minutes:02}' + ":" + f'{self.__seconds:02}' + " " + self.__am_or_pm

    def tick (self):
        self.__seconds += 1

        if self.__seconds == 60:
            self.__seconds = 0
            self.__minutes += 1

            if self.__minutes == 60:
                self.__minutes = 0
                self.__hours += 1

                if self.__hours == 12:
                    if self.__am_or_pm == "AM":
                        self.__am_or_pm = "PM"
                    else:
                        self.__am_or_pm = "AM"

                elif self.__hours == 13:
                    self.__hours = 1


def main ():
    clock = Clock (11, 59, 59, "PM")
    print (clock)
    clock.tick ()
    print (clock)

    clock.advance_time (15)
    print (clock)

if __name__ == "__main__":
    main ()

File: Unit02/Clock.java
public class Clock {
    private int hours;
    private int minutes;
    private int seconds;
    private String amOrPm;

    Clock (int hours, int minutes, int seconds, String amOrPm) {
        this.hours = hours;
        this.minutes = minutes;
        this.seconds = seconds;
        this.amOrPm = amOrPm;
    }

    public int getHours () {
        return this.hours;
    }

    public int getMinutes () {
        return this.minutes;
    }

    public int getSeconds () {
        return this.seconds;
    }

    public String getAmOrPM () {
        return this.amOrPm;
    }

    @Override
    public String toString () {
        return this.hours +  ":" + this.minutes + ":" + this.seconds + ":" + this.amOrPm;
    }

    public void tick() {
        this.seconds += 1;
    
        if(this.seconds == 60) {
            this.seconds = 0;
            this.minutes += 1;
    
            if(this.minutes == 60) {
                this.minutes = 0;
                this.hours += 1;
        
                if(this.hours == 12) {
                    if(this.amOrPm.equals("AM")) {
                        this.amOrPm = "PM";
                    } else {
                        this.amOrPm = "AM";
                    }
                } else if(this.hours == 13) {
                    this.hours = 1;
                }
            }
        }
    }
    
    public void advanceTime (int seconds) {
        while (seconds > 0) {
            tick ();
            System.out.println(this);
            seconds = seconds - 1;
        }
    }

    public static void main(String [] arg) {
        Clock clock = new Clock (11, 59, 59, "PM");
        System.out.println (clock);
        clock.tick();
        System.out.println(clock);
        clock.advanceTime(10);

    }
}
